src/Build.UnitTests/ProjectCache/ProjectCacheTests.cs

```diff
@@ -807,57 +807,157 @@ public static IEnumerable<object[]> CacheExceptionLocationsTestData
+                var exceptionsThatEndUpInBuildResult =
+                    ExceptionLocations.Constructor | ExceptionLocations.BeginBuildAsync | ExceptionLocations.GetCacheResultAsync;
+
+                if ((exceptionsThatEndUpInBuildResult & exceptionLocations) != 0)
+                {
+                    buildResult.OverallResult.ShouldBe(BuildResultCode.Failure);
+                    buildResult.Exception.Message.ShouldContain("Cache plugin exception from");
+                }
+
+                // BuildManager.EndBuild calls plugin.EndBuild, so if only plugin.EndBuild fails it means everything else passed,
+                // so the build submission should be successful.
+                if (exceptionLocations == ExceptionLocations.EndBuildAsync)
+                {
+                    buildResult.OverallResult.ShouldBe(BuildResultCode.Success);
+                }
+            }
+            finally
+            {
+                // These exceptions prevent the creation of a plugin so there's no plugin to shutdown.
+                var exceptionsThatPreventEndBuildFromThrowing = ExceptionLocations.Constructor |
+                                                                ExceptionLocations.BeginBuildAsync;
+
+                if ((exceptionLocations & exceptionsThatPreventEndBuildFromThrowing) != 0 ||
+                    !exceptionLocations.HasFlag(ExceptionLocations.EndBuildAsync))
+                {
+                    Should.NotThrow(() => buildSession!.Dispose());
+                }
+                else if (exceptionLocations.HasFlag(ExceptionLocations.EndBuildAsync))
+                {
+                    var e = Should.Throw<Exception>(() => buildSession!.Dispose());
+                    e.Message.ShouldContain("Cache plugin exception from EndBuildAsync");
+                }
+                else
+                {
+                    throw new NotImplementedException();
+                }
+            }
+
+            logger.BuildFinishedEvents.First().Succeeded.ShouldBeFalse();
+
+            // Plugin query must happen after plugin init. So if plugin init fails, then the plugin should not get queried.
+            var exceptionsThatShouldPreventCacheQueryAndEndBuildAsync = ExceptionLocations.Constructor | ExceptionLocations.BeginBuildAsync;
+
+            if ((exceptionsThatShouldPreventCacheQueryAndEndBuildAsync & exceptionLocations) != 0)
+            {
+                logger.FullLog.ShouldNotContain($"{AssemblyMockCache}: GetCacheResultAsync for");
+                logger.FullLog.ShouldNotContain($"{AssemblyMockCache}: EndBuildAsync");
+            }
+            else
+            {
+                StringShouldContainSubstring(logger.FullLog, $"{AssemblyMockCache}: GetCacheResultAsync for", expectedOccurrences: 1);
+                StringShouldContainSubstring(logger.FullLog, $"{AssemblyMockCache}: EndBuildAsync", expectedOccurrences: 1);
+            }
+
+            // TODO: this ain't right now is it?
```

Comment:
To be done in the follow-up PR, or is this leftover? 🤔

Reply:
Yup, followup-pr: #6359
This PR just fixes the exception handling part, follow-up PR does the logging.